services/src/main/scala/cromwell/services/healthmonitor/HealthMonitorServiceActor.scala

```diff
@@ -0,0 +1,119 @@
+    result.withTimeout(futureTimeout, errMsg, context.system.scheduler)
+      .recover { case NonFatal(ex) =>
+        failedStatus(ex.getMessage)
+      } map {
+      Store(subsystem, _)
+    } pipeTo self
+
+    ()
+  }
+
+  private def store(subsystem: MonitoredSubsystem, status: SubsystemStatus): Unit = {
+    data = data + ((subsystem, CachedSubsystemStatus(status, System.currentTimeMillis)))
+    logger.debug(s"New health monitor state: $data")
+  }
+
+  private def getCurrentStatus: StatusCheckResponse = {
+    val now = System.currentTimeMillis()
+    // Convert any expired statuses to unknown
+    val processed = data map {
+      case (s, c) if now - c.created > staleThreshold.toMillis => s.name -> UnknownStatus
+      case (s, c) => s.name -> c.status
+    }
+
+    // overall status is ok iff all subsystems are ok
+    val overall = processed.forall(_._2.ok)
+    
+    StatusCheckResponse(overall, processed)
+  }
+}
+
+object HealthMonitorServiceActor {
+  val DefaultFutureTimeout: FiniteDuration = 1 minute
+  val DefaultStaleThreshold: FiniteDuration = 15 minutes
+  val DefaultSweepTime: FiniteDuration = 1 minute // 5 minutes
+
+  private case object CheckTickKey
+
+  val OkStatus = SubsystemStatus(true, None)
+  val UnknownStatus = SubsystemStatus(false, Option(List("Unknown status")))
+  def failedStatus(message: String) = SubsystemStatus(false, Option(List(message)))
+
+  final case class MonitoredSubsystem(name: String, check: Unit => Future[SubsystemStatus])
+  final case class SubsystemStatus(ok: Boolean, messages: Option[List[String]])
+  final case class CachedSubsystemStatus(status: SubsystemStatus, created: Long) // created is time in millis when status was captured
+
+  sealed abstract class HealthMonitorServiceActorRequest
+  case object CheckAll extends HealthMonitorServiceActorRequest
+  final case class Store(subsystem: MonitoredSubsystem, status: SubsystemStatus) extends HealthMonitorServiceActorRequest
+  case object GetCurrentStatus extends HealthMonitorServiceActorRequest with ServiceRegistryMessage { override val serviceName = "HealthMonitor" }
+
+  sealed abstract class HealthMonitorServiceActorResponse
+  final case class StatusCheckResponse(ok: Boolean, systems: Map[String, SubsystemStatus]) extends HealthMonitorServiceActorResponse
+
+  /**
+    * Adds non-blocking timeout support to futures.
+    * Example usage:
+    * {{{
+    *   val future = Future(Thread.sleep(1000*60*60*24*365)) // 1 year
+    *   Await.result(future.withTimeout(5 seconds, "Timed out"), 365 days)
+    *   // returns in 5 seconds
+    * }}}
+    */
+  private implicit class FutureWithTimeout[A](f: Future[A]) {
+    def withTimeout(duration: FiniteDuration, errMsg: String, scheduler: Scheduler)(implicit ec: ExecutionContext): Future[A] =
+      Future.firstCompletedOf(List(f, after(duration, scheduler)(Future.failed(new TimeoutException(errMsg)))))
```

Comment:
I think this might not be working right. Thanks to my docker being hosed my docker commands always hang and yet I'm always getting Unknown instead of the timeout

Reply:
Also as a note, checking status of docker by pulling an image means that a docker daemon needs to be running and accessible to wherever Cromwell is running, which might not be a given if Cromwell itself is running in a docker..

Reply:
Spoke to @Horneth in person - going to try to pull in the non-local docker client and use that to ping dockerhub instead. in some ways that'll be better, in other ways worse, but it'd at least work :)